refactoring.md

```diff
@@ -0,0 +1,95 @@
+# Refactoring
+
+## Why do you need refactoring ?
+
+Refactoring your codebase terraform is a necessity within the lifecycle of your codebase.
+As your infrastructure grows and as you add collaborators to the project, you'll need to reconsider things like :
+
+- the size of the layers
+- the scope of the modules
+- the interface of the modules
+
+To keep track of your continuous changes, you should set up "quality probes", which will serve as goals the refactored codebase aims to reach like :
+
+- the maximum time a plan can take
+- the number of variables you need to provide before implementing the layer
+- the time it takes to implement your 3 most frequent tasks.
+
+## Questions to ask yourself
+
+Your priorities regarding a refactoring are the following in this order :
+
+1. Control code quality
+2. Accelerate feature promotion
+3. Reduce code duplication
+4. Improve readability
+
+![refacto_decision_tree](assets/img/refacto_decision_tree.png)
+
+### How many collaborators are contributing to the codebase ? #CodeQualityControl
+
+- Less than 2 *aka you and a partner*.
+
+  You should frequently work in pair and communicate on the refacto. You must review each others PR, that way both of you will have the entire knowledge of what's merged into the codebase.
+
+- 3 collaborators and more *aka a team*.
+
+  What's risky with a team refactoring a codebase is poor quality code merged into the codebase due to a loss in information transmission. To control code quality over time during the refactoring you need to proceed as follows :
+
+  - Write out the new standards you want to implement
+  - Dedicate time on formation for members of the team on the new standards
+  - Archive your current codebase
```

Comment:
@Alan-pad I mitigate this part with [recommendations](https://github.com/padok-team/docs-terraform-guidelines/pull/19/files#diff-8fa2124a4b79ce2dfe8e574d82b0f2badf6e490302709a47f9c75701981276ecR41-R45) instead of this hard recommendation
> *Archive your current codebase*